ast/io/ast_reader.c

```diff
@@ -0,0 +1,856 @@
+struct Type* readType(FILE* file, bool debug){
+	
+	if(debug){ printf("readType(...)\n"); }
+	
+	magic_num_require(MAGIC_TYPE, file);
+	
+	const int kind = deserialize_int(file);;
+	
+	struct Type* b = malloc(sizeof(struct Type));
+	
+	b->m1 = NULL;
+	b->m2 = NULL;
+	b->m3 = NULL;
+
+	switch(kind){
+		case 1: b->m1 = readBasicTypeWrapped(file, debug); break;
+		case 2: b->m2 = readTypeParam(file, debug); break;
+		case 3: b->m3 = readArrayType(file, debug); break;
+		default:
+			printf("Error in readType\n");
+			free(b);
+			fclose(file);
+			exit(1);
+	}
+	
+	magic_num_require(MAGIC_END_TYPE, file);
+	
+	return b;
+}
+struct SubrType* readSubrType(FILE* file, bool debug){
+	
+	if(debug){ printf("readSubrType(...)\n"); }
+	
+	magic_num_require(MAGIC_SUBRTYPE, file);
+	
+	struct SubrType* v = 
+		malloc(sizeof(struct SubrType));
+
+	v->returnType = readType(file, debug);
+	
+	v->hasSideEffects = deserialize_int(file);
+	v->count_argTypes = deserialize_int(file);
+	
+	v->argTypes = 
+		malloc(sizeof(struct Type*)*(v->count_argTypes));
+	for(int i=0;i < (v->count_argTypes); i++){
+		v->argTypes[i] = readType(file, debug);
+	}
+	
+	magic_num_require(MAGIC_END_SUBRTYPE, file);
+	
+	return v;
+}
+struct SimpleType* readSimpleType(FILE* file, bool debug){
+	
+	if(debug){ printf("readSimpleType(...)\n"); }
+	
+	magic_num_require(MAGIC_SIMPLETYPE, file);
+	
+	struct SimpleType* v = malloc(sizeof(struct SimpleType));
+	
+	char* tmp = deserialize_string(file);
+	strcpy(v->typeName, tmp);
+	free(tmp);
+	
+	v->isPrimitive = deserialize_int(file);
+	v->isIntType   = deserialize_int(file);
+	
+	v->typeParamCount = deserialize_int(file);
+	
+	if(v->typeParamCount > 0){
+		v->typeParams = malloc(sizeof(uint8_t)*(v->typeParamCount));
+	}
+	
+	for(int i=0;i<v->typeParamCount;i++){
+		
+		v->typeParams[i] = deserialize_int(file);
+	}
+	
+	magic_num_require(MAGIC_END_SIMPLETYPE, file);
+	
+	return v;
+}
+struct ArrayType* readArrayType(FILE* file, bool debug){
+	
+	if(debug){ printf("readArrayType(...)\n"); }
+	
+	magic_num_require(MAGIC_ARRAYTYPE, file);
+	
+	struct ArrayType* v = malloc(sizeof(struct ArrayType));
+	
+	v->element_type = readType(file, debug);
+	
+	magic_num_require(MAGIC_END_ARRAYTYPE, file);
+	
+	return v;
+}
+struct TypeParam* readTypeParam(FILE* file, bool debug){
+	
+	if(debug){ printf("readTypeParam(...)\n"); }
+	
+	magic_num_require(MAGIC_TYPEPARAM, file);
+	
+	struct TypeParam* v = malloc(sizeof(struct TypeParam));
+		
+	fread(v, sizeof(struct TypeParam), 1, file);
+	
+	magic_num_require(MAGIC_END_TYPEPARAM, file);
+	
+	return v;
+}
+struct BasicTypeWrapped* readBasicTypeWrapped(FILE* file, bool debug){
+	
+	if(debug){ printf("readBasicTypeWrapped(...)\n"); }
+	
+	magic_num_require(MAGIC_BASICTYPEWRAPPED, file);
+	
+	const int kind = deserialize_int(file);
+	
+	struct BasicTypeWrapped* v = malloc(sizeof(struct BasicTypeWrapped));
+	
+	switch(kind){
+		case 1: 
+			v->simpleType = readSimpleType(file, debug);
```

Comment:
maybe initialize both simpleType and subrType to NULL and therefore less code in  the indented section